chain/manager.go

```diff
@@ -298,6 +301,55 @@ func (m *Manager) AddBlocks(blocks []types.Block) error {
 	return nil
 }
 
+// AddValidatedV2Blocks ingests a chain of v2 blocks. The blocks must already be
+// validated, and the first block's parent must be known. If the chain has
+// sufficient work, it may become the new best chain, triggering a reorg.
+func (m *Manager) AddValidatedV2Blocks(blocks []types.Block, states []consensus.State) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if len(blocks) == 0 {
+		return nil
+	}
```

Comment:
should probably compare the length of `blocks` and `states` here to avoid panics further down.